### PR TITLE
Add and use wistd::invoke

### LINKED ISSUE #36: std::invoke for callables.
C++17 allows some pretty crazy things with templates. Observe the following:

```cpp
template <typename T>
struct member_parent { };

template<typename Object, typename Return, typename... Args>
struct member_parent<Return(Object:: *)(Args...)> {
	typedef Object type;
};

template<typename T>
using member_parent_t = typename member_parent<T>::type;

template<auto close_fn, typename token_t, token_t invalid_token = token_t()>
using com_token = wil::unique_com_token<member_parent_t<decltype(close_fn)>, token_t, decltype(close_fn), close_fn, invalid_token>;
```

This code allows to simplify writing a custom COM token:
```cpp
using app_visibility_sink_cookie = com_token<&IAppVisibility::Unadvise, DWORD>;
```

However, it can't be easily included in wil without depending on C++17 (we can't overload templated type aliases, C++ determined that would be too easy).

The code should work, but doesn't, because of this line: https://github.com/microsoft/wil/blob/ade910fb44f0b23c03111a2838c2148d326f5622/include/wil/resource.h#L1370

If that line where to be changed to `std::invoke(close_fn, m_source, m_token);`, this would work. However, I am aware that wil tries to avoid dependencies on the system standard library, and I am honestly unsure on how I would proceed to add more stuff to `wistd`.

Ideally, replacing all of the calls of function pointer template parameters like above by calls to `wistd::invoke` would allow more creativity in the use of wil.

### PR DESCRIPTION
**Pull request by sylveon**
Fixes #36

### CODE REVIEW

File: include/wil/wistd_functional.h [hunk 1]
```diff
@@ -516,9 +516,16 @@ namespace wistd     // ("Windows Implementation" std)
     swap_wil(function<_Rp(_ArgTypes...)>& __x, function<_Rp(_ArgTypes...)>& __y)
     {return __x.swap(__y);}
 
+    // std::invoke
```
Comment (dunhor) [inline include/wil/wistd_functional.h]: Please copy the implementation from libc++

Comment (sylveon) [reply include/wil/wistd_functional.h]: The libc++ implementation uses invoke_result_t, which requires C++14.

Comment (dunhor) [reply include/wil/wistd_functional.h]: Feel free to reduce that limitation as there's no requirement that `invoke_result` be limited to C++17, apart from that's when it was introduced to the standard. Could also use `__invoke_of`

Comment (sylveon) [reply include/wil/wistd_functional.h]: Fixed

**Comment by ChrisGuzak** [comment]
update the description, "this set wil up to support [callables ](https://en.cppreference.com/w/cpp/named_req/Callable) instead of just functions in the resource cleanup"

**Comment by sylveon** [comment]
Description of what?

**Comment by ChrisGuzak** [comment]
the PR description.

**Comment by ChrisGuzak** [comment]
I'm not github PR savvy, is there a description independent of the title?

**Comment by sylveon** [comment]
Ah yes there is. Kinda pointless to change the description since it's merged now but I updated it anyways.

**Pull request merged**